### ISSUE TITLE
[0023] `auto` keyword, at least C++11 style

### ISSUE THREAD

**Issue by devshgraphicsprogramming** [microsoft/hlsl-specs#24]
**Is your feature request related to a problem? Please describe.**

The lack of `auto` makes it impossible to define the following very useful function
```cpp
template<typename DfDu_t, typename DuDx_t>
auto applyChainRule(const DfDu_t DfDu, const DuDx_t DuDx)
{
    return DfDu*DuDx;
}
```
as it stands right now I have to provide a manual overload or specialization for every `floatNxP` `floatPxM` combo.

Also because there are no constructors, the code gets really long and redundant.
```cpp
template<class A, class B, class C>
struct MyReallyVeryLongNamedStruct
{
    static MyReallyVeryLongNamedStruct<A,B,C> create(); 
};

void main()
{
    MyReallyVeryLongNamedStruct<SomethingA,AnotherB,OtherC> annoyingVariableToShowcase = MyReallyVeryLongNamedStruct<SomethingA,AnotherB,OtherC>::create();
}
```

**Describe the solution you'd like**

Being able to use `auto` for when:
- return type
- variable type due to assignment
can already be deduced trivially.

**Describe alternatives you've considered**

I could make an ugly `template struct` with and internal `using` alias which tells me the resulting type of every matrix multiplication.
```cpp
template<typename DfDu_t, typename DuDx_t>
impl::matrix_mul<DfDu_t,DuDx_t>::result_t applyChainRule(const DfDu_t DfDu, const DuDx_t DuDx)
{
    return DfDu*DuDx;
}
```
But that shifts the problem of having to do the manual specialization onto a struct.

Also that doesn't scale well to other things.

**Comment by devshgraphicsprogramming** [comment]
fun fact, we also tried
```cpp
template<typename DfDx_t, typename DfDu_t, typename DuDx_t>
DfDx_tapplyChainRule(const DfDu_t DfDu, const DuDx_t DuDx)
{
    return DfDu*DuDx;
}
```
but its beyond DXC's capability to deduce that for:
```cpp
float2x3 dFdG;
float3x4 dGdR;
float2x4 dFdR = applyChainRule(dFdG, dGdR);
```

**Comment by llvm-beanz** [comment]
Yea... C++ Templates don't infer types the way you get with generics in other languages. The template's type needs to be derivable from the instantiation site unless the return type is `auto`.

I will also take this one. I am confident that eventually we will want to add `auto`.

**Comment by tex3d** [comment]
Just FYI: you can use the `matrix` template to implement your template function without the use of `auto`.  Note that you should be using the `mul()` intrinsic for matrix multiplication instead of the `*` operator which always does a component-wise multiplication.
```hlsl
template<typename C, int N, int M, int P>
matrix<C, N, M> applyChainRule(matrix<C, N, P> DfDu, matrix<C, P, M> DuDx) {
  return mul(DfDu, DuDx);
}
```

**Comment by llvm-beanz** [comment]
HLSL 2021 added templates and `using` for namespaces and type aliases (including template type aliases).

The matrix template is a built-in template and is present in all HLSL modes supported by DXC. The non-template matrix types are just type aliases of the matrix type.

**Comment by devshgraphicsprogramming** [comment]
> HLSL 2021 added templates and `using` for namespaces and type aliases (including template type aliases).
> 
> The matrix template is a built-in template and is present in all HLSL modes supported by DXC. The non-template matrix types are just type aliases of the matrix type.

are vector types similarly templated?

**Comment by llvm-beanz** [comment]
> are vector types similarly templated?

Yep. Vectors are `vector<T, N>`, and that does work with HLSL 2021 template parameters.

**Comment by llvm-beanz** [comment]
This should be part of [proposal 23](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0023-cxx11-base.md)
